WordPress/gutenberg
# Add new ariakit-based `DropdownMenu` component

Part of #50459
Requires #55365

## What?


Add a new version of the `DropdownMenu` component based off the `ariakit` library.

**_Note: The component should still be considered a WIP and will be tweaked and refined either in this PR or via follow-up PRs as it receives feedback and as it gets used throughout the codebase_**

## Why?


We found a few blockers when working on the radix-ui based new version of `DropdownMenu`, and therefore we're trying a version based on `ariakit` instead

## How?


- followed examples from ariakit's website
- requirements, styles, stories and unit tests were heavily inspired from the radix-ui version of the experimental component

#### Next steps:

 - Refine design and behaviour
 - Address any bugs
 - Start testing in the editor
 - Try to write a compat layer for legacy `DropdownMenu` and `MenuItem`
 
<details>

<summary>Progress tracker</summary>

- Prep basic version of components:
  - [x] `DropdownMenu`
    - [x] controlled mode
    - [x] trigger (use `DropdownMenuItem`)
    - [x] children
  - [x] `DropdownMenuGroup`
  - [x] `DropdownMenuGroupLabel` (or `DropdownMenuLabel`)
  - [x] `DropdownMenuItem`
      - [x] disabled
      - [x] onClick
      - [x] hideOnClick
      - [x] prefix and suffix
  - [x] `DropdownMenuCheckboxItem`
      - [x] controlled/uncontrolled (checked/onChange/defaultChecked..)
      - [x] multiple values
      - [x] same as menu item, but no prefix (taken by check icon)
  - [x] `DropdownMenuRadioItem`
      - [x] controlled/uncontrolled (checked/onChange/defaultChecked..)
      - [x] same as menu item, but no prefix (taken by radio icon)
  - [x] `DropdownMenuSeparator` 
- [x] Add back comprehensive Storybook examples
- [x] Tweak styles to meet design spec
- [x] Add animation (either via CSS or via framer motion)
- [x] Check reduced motion support
- [x] Check RTL support:
  - [x] it should flip left/right placements
  - [x] it should affect keyboard navigation
  - [x] set the `dir` attribute via HTML
- [x] Support toolbar variant via components context
- [x] edge case: dropdown opened inside a Modal, we should use `hideOnEscape={ ( e ) => { e.stopPropagation(); return true; } }`
- [x] Export via private APIs
- [x] Add back unit tests
- [x] Add types JSDocs
- [x] Add README
- [x] Check that all actionable TODOs in the code have been done, open follow-up issues for non-solvable items
</details>

## Testing Instructions





1. `npm run distclean && npm ci && npm run storybook:dev`
2. Load the `DropdownMenuV2Ariakit` storybook examples
3. Interact with the example, and make sure that the dropdown:
    - uses semantically correct markup and attributes/roles
    - behaves as expected
    - looks as expected (including high contrast mode)
    - doesn't have any clear missing feature

## Screenshots or screencast 

_**(ignore the fact that menu's popover is not aligned with the trigger when Storybook's sidebar is opened, as that's a `floating-ui` bug not related to this PR)**_

https://github.com/WordPress/gutenberg/assets/1083581/685beb1b-3107-422c-82d5-4f48eb0df7b3
